### PR TITLE
perf: use rclonefile to fix cloning on macOS

### LINKED ISSUE #5001: Fix file cloning on macOS
pnpm currently uses hard links (not cloning) on macOS. This is due to an issue in Node.js: https://github.com/libuv/libuv/pull/2578

As a result, pnpm is much slower on macOS. Looks like [bun](https://bun.sh/) doesn't have this issue and as a result is a lot faster than pnpm in macOS. If Node.js is refusing to merge a fix, maybe there's another way to work it out.

Contributors are welcomed as I don't have a mac and probably won't have till the war ends.

### PR DESCRIPTION
**Pull request by zkochan**
close #5001

### CODE REVIEW

**Comment by weyert** [comment]
Why would it try to installation `rclonefile` native library for non macos platforms?

**Comment by weyert** [comment]
If I understand it correctly the `index.js` (https://github.com/sverrejoh/rclonefile/blob/main/index.js) is wrong as it also defines windows/linux builds which are suggested not to be created/build as part of the package

**Comment by zkochan** [comment]
I know. I only did the PR to not forget about it.

**Comment by RexSkz** [comment]
I wonder if the PR has an expected finish time, or is it still on the TODO list?

I tried the latest Node.js 20.6.1 with `package-import-method = clone` on my MacOS (Ventura 13.5.1) and found it still throws the error `ENOSYS: function not implemented`, which means the Node.js bug hasn't been fixed, this PR is still worth to move on.

The issue with `rclonefile` may be solved by calling it only when `process.platform === 'darwin'`.

File: fs/indexed-pkg-importer/src/index.ts [hunk 1]
```diff
@@ -1,9 +1,11 @@
-import { constants, type Stats, existsSync } from 'fs'
+import { type Stats, existsSync } from 'fs'
+import { constants } from 'fs/promises'
```
Comment (zkochan) [inline fs/indexed-pkg-importer/src/index.ts]: why was this change needed?

Comment (nachoaldamav) [reply fs/indexed-pkg-importer/src/index.ts]: Oh, in the 1st commit you removed `constants` and I added it too, I'll import it from `fs` so it's one line only

**Comment by zkochan** [comment]
This change makes install faster almost twice on macOS, when the lockfile is up to date and the store has all the packages.

**Comment by zkochan** [comment]
Unfortunately we'll have to ship all the refclone artefacts with the plain js version of pnpm (this will increase the size of the tarball by 1.3MB). If don't do that, corepack won't be able to install pnpm with reflink support.

**Comment by gluxon** [comment]
To add another data point, this improved performance significantly on a real world codebase with 1000+ workspace packages and 5000+ external dependencies.

|Commit|Trial 1|Trial 2|Trial 3|Average|
|-|-|-|-|-|
|12f45a83d2cfaba166f4eda82e573736633900b5 (Parent commit)|3m 15.7s|3m 10.7s|3m26.2|3m 17.5s|
|4d18bec33c2ee5f3f178e1e84a8ff74d5ff968e4 (This PR)|2m 31.3s|2m 19.1s|2m12.8s|2m 21.1s|

**That's about 40% faster on this giant codebase.** What a great change. 🥳

**Comment by intrnl** [comment]
the gating of this functionality to Windows and macOS does mean that Linux users using filesystems such as BTRFS (default on Fedora and OpenSUSE distributions) and XFS don't get to benefit from this, no?

**Pull request merged**